### PR TITLE
HIVE-22997 : Copy external table to target during Repl Dump operation

### PR DESCRIPTION
**Pull request by pkumarsinha**

### CODE REVIEW

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationOnHDFSEncryptedZones.java [hunk 1]
```diff
@@ -102,12 +104,20 @@ public void targetAndSourceHaveDifferentEncryptionZoneKeys() throws Throwable {
           put(HiveConf.ConfVars.REPLDIR.varname, primary.repldDir);
         }}, "test_key123");
 
+    List<String> dumpWithClause = Arrays.asList(
```
Comment (aasha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationOnHDFSEncryptedZones.java]: Why is this needed?

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java [hunk 2]
```diff
@@ -397,17 +397,19 @@ public void externalTableWithPartitions() throws Throwable {
...
     assertTablePartitionLocation(primaryDbName + ".t2", replicatedDbName + ".t2");
   }
 
   @Test
   public void externalTableIncrementalReplication() throws Throwable {
-    WarehouseInstance.Tuple tuple = primary.dumpWithCommand("repl dump " + primaryDbName);
+    List<String> withClause = externalTableBasePathWithClause();
+    String replDumpCommand = "repl dump " + primaryDbName
+            + " WITH (" + withClause.get(0) + "," + withClause.get(1) + ")";
+    WarehouseInstance.Tuple tuple = primary.dumpWithCommand(replDumpCommand);
```
Comment (aasha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java]: can use withClause directly

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java [hunk 3]
```diff
@@ -436,16 +438,29 @@ public void externalTableIncrementalReplication() throws Throwable {
...
-    replica.load(replicatedDbName, primaryDbName, loadWithClause)
+    replica.load(replicatedDbName, primaryDbName, withClause)
         .run("use " + replicatedDbName)
         .run("show tables like 't1'")
         .verifyResult("t1")
         .run("show partitions t1")
         .verifyResults(new String[] { "country=india", "country=us" })
         .run("select place from t1 order by place")
-        .verifyResults(new String[] { "bangalore", "mumbai", "pune" })
+        .verifyResults(new String[] {})
```
Comment (aasha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java]: Why is there no data here?

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java [hunk 4]
```diff
@@ -713,9 +732,11 @@ public void testExternalTableDataPath() throws Exception {
 
   @Test
   public void testExternalTablesIncReplicationWithConcurrentDropTable() throws Throwable {
-    List<String> dumpWithClause = Collections.singletonList(
-            "'" + HiveConf.ConfVars.REPL_INCLUDE_EXTERNAL_TABLES.varname + "'='true'"
-    );
+    List<String> dumpWithClause = Arrays.asList(
```
Comment (aasha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java]: Can be separated in a common method

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestTableLevelReplicationScenarios.java [hunk 5]
```diff
@@ -918,7 +922,9 @@ public void testRenameTableScenariosWithReplaceExternalTable() throws Throwable
     String newPolicy = primaryDbName + ".'(in[0-9]+)|(out1500)|(in2)'";
     dumpWithClause = Arrays.asList(
             "'" + HiveConf.ConfVars.REPL_INCLUDE_EXTERNAL_TABLES.varname + "'='true'",
-            "'" + HiveConf.ConfVars.REPL_BOOTSTRAP_EXTERNAL_TABLES.varname + "'='false'"
+            "'" + HiveConf.ConfVars.REPL_BOOTSTRAP_EXTERNAL_TABLES.varname + "'='false'",
```
Comment (aasha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestTableLevelReplicationScenarios.java]: Can be added to a separate method

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java [hunk 6]
```diff
@@ -122,25 +127,31 @@ public String getName() {
...
-      Path hiveDumpRoot = new Path(currentDumpPath, ReplUtils.REPL_HIVE_BASE_DIR);
-      DumpMetaData dmd = new DumpMetaData(hiveDumpRoot, conf);
-      // Initialize ReplChangeManager instance since we will require it to encode file URI.
-      ReplChangeManager.getInstance(conf);
-      Path cmRoot = new Path(conf.getVar(HiveConf.ConfVars.REPLCMDIR));
-      Long lastReplId;
-      if (!dumpRoot.getFileSystem(conf).exists(dumpRoot)
-              || dumpRoot.getFileSystem(conf).listStatus(dumpRoot).length == 0) {
-        lastReplId = bootStrapDump(hiveDumpRoot, dmd, cmRoot, hiveDb);
+      //First Check if external table copy work has been initialized, if so, just do that and return.
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java]: Please add more comments here

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java [hunk 7]
```diff
@@ -367,13 +379,14 @@ private Long incrementalDump(Path dumpRoot, DumpMetaData dmd, Path cmRoot, Hive
...
                   && shouldDumpExternalTableLocation()) {
-              writer.dataLocationDump(table);
+              extTableLocations.addAll(writer.dataLocationDump(table));
             }
 
             // Dump the table to be bootstrapped if required.
             if (shouldBootstrapDumpTable(table)) {
               HiveWrapper.Tuple<Table> tableTuple = new HiveWrapper(hiveDb, dbName).table(table);
-              dumpTable(dbName, tableName, validTxnList, dbRoot, dumpRoot, bootDumpBeginReplId, hiveDb, tableTuple);
+              replPathMappings.addAll(dumpTable(dbName, tableName, validTxnList, dbRoot, dumpRoot, bootDumpBeginReplId,
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java]: If there are too many external tables OOM?

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationOnHDFSEncryptedZones.java (hunk 1 above)
Comment (pkumarsinha) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationOnHDFSEncryptedZones.java]: Since the external table copy is happening on source, the configs are required to be passed at dump.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java (hunk 3 above)
Comment (pkumarsinha) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java]: Since the modification happened after dump, data will be seen only after next dump-load cycle, which is being tested.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java (hunk 6 above)
Comment (pkumarsinha) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java]: Will do

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java (hunk 7 above)
Comment (pkumarsinha) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java]: That's the existing issue with the way current iterators are being used and the patch doesn't address that. May be later, we may want to solve that in general. May be by using some disk based iterators. Will file a tracking JIRA for that.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestTableLevelReplicationScenarios.java (hunk 5 above)
Comment (pkumarsinha) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestTableLevelReplicationScenarios.java]: Will see if we can do that. Problem was that the parameters were not fixed and were being set on case-to-case basis.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java [hunk 8]
```diff
@@ -386,8 +399,10 @@ private Long incrementalDump(Path dumpRoot, DumpMetaData dmd, Path cmRoot, Hive
         }
       }
       dumpTableListToDumpLocation(tableList, dumpRoot, dbName, conf);
+      List<ExternalTableCopyTaskBuilder.DirCopyWork> extTableCopyWorks = dirLocationsToCopy(extTableLocations);
```
[the remark this answers is not in the document]
Comment (pkumarsinha) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java]: Sure, will fix that.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java [hunk 9]
```diff
@@ -122,25 +127,31 @@ public String getName() {
...
-      // Initialize ReplChangeManager instance since we will require it to encode file URI.
-      ReplChangeManager.getInstance(conf);
-      Path cmRoot = new Path(conf.getVar(HiveConf.ConfVars.REPLCMDIR));
-      Long lastReplId;
-      if (!dumpRoot.getFileSystem(conf).exists(dumpRoot)
-              || dumpRoot.getFileSystem(conf).listStatus(dumpRoot).length == 0) {
-        lastReplId = bootStrapDump(hiveDumpRoot, dmd, cmRoot, hiveDb);
+      //First Check if external table copy work has been initialized, if so, just do that and return.
+      if (work.dirCopyIteratorInitialized() || work.replPathIteratorInitialized()) {
+        intitiateDataCopyTasks();
```
[the remark this answers is not in the document]
Comment (pkumarsinha) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java]: We would like to have a control on how many tasks in parallel can run including external table copy. That was the existing behavior even in case of external table copy.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationOnHDFSEncryptedZones.java [hunk 10]
```diff
@@ -102,12 +104,20 @@ public void targetAndSourceHaveDifferentEncryptionZoneKeys() throws Throwable {
...
+            "'hive.repl.add.raw.reserved.namespace'='true'",
+            "'" + HiveConf.ConfVars.REPL_EXTERNAL_TABLE_BASE_DIR.varname + "'='"
+                    + replica.externalTableWarehouseRoot + "'",
+            "'distcp.options.skipcrccheck'=''",
+            "'" + HiveConf.ConfVars.HIVE_SERVER2_ENABLE_DOAS.varname + "'='false'",
+            "'" + HiveConf.ConfVars.HIVE_DISTCP_DOAS_USER.varname + "'='"
+                    + UserGroupInformation.getCurrentUser().getUserName() +"'");
```
Comment (aasha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationOnHDFSEncryptedZones.java]: Please add a test with scheduler as well

Comment (pkumarsinha) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationOnHDFSEncryptedZones.java]: Done

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java (hunk 4 above)
Comment (pkumarsinha) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java]: Done

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestTableLevelReplicationScenarios.java (hunk 5 above)
Comment (pkumarsinha) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestTableLevelReplicationScenarios.java]: Done

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java (hunk 3 above)
Comment (pkumarsinha) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java]: The copy was happening in repl load, change was prior to that.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestTableLevelReplicationScenarios.java (hunk 5 above)
Comment (pkumarsinha) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestTableLevelReplicationScenarios.java]: Yes, I am not planning to refactor all the places.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationOnHDFSEncryptedZones.java [hunk 11]
```diff
@@ -102,12 +104,20 @@ public void targetAndSourceHaveDifferentEncryptionZoneKeys() throws Throwable {
           put(HiveConf.ConfVars.REPLDIR.varname, primary.repldDir);
         }}, "test_key123");
 
+    List<String> dumpWithClause = Arrays.asList(
+            "'hive.repl.add.raw.reserved.namespace'='true'",
+            "'" + HiveConf.ConfVars.REPL_EXTERNAL_TABLE_BASE_DIR.varname + "'='"
+                    + replica.externalTableWarehouseRoot + "'",
+            "'distcp.options.skipcrccheck'=''",
```
[the remark this answers is not in the document]
Comment (pkumarsinha) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationOnHDFSEncryptedZones.java]: That was part of existing test, I have used the same set of config in dump time

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java (hunk 3 above)
Comment (pkumarsinha) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java]: Because copy is happening during load.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java [hunk 12]
```diff
@@ -662,17 +690,48 @@ void dumpTable(String dbName, String tblName, String validTxnList, Path dbRoot,
...
-      LOG.info("Scheduled a repl copy task from [{}] to [{}]",
-              replPathMapping.getSrcPath(), replPathMapping.getTargetPath());
+    if (!childTasks.isEmpty()) {
+      boolean ackTaskAdded = false;
+      if (taskTracker.canAddMoreTasks()) {
+        childTasks.add(dumpCompleteAckWorkTask);
+        ackTaskAdded = true;
+      }
+      if (hasMoreCopyWork() || !ackTaskAdded) {
+        DAGTraversal.traverse(childTasks, new AddDependencyToLeaves(TaskFactory.get(work, conf)));
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java]: Add the ack only after no more work is pending

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java [hunk 13]
```diff
@@ -78,22 +81,26 @@
...
 import java.util.concurrent.TimeUnit;
 import static org.apache.hadoop.hive.ql.exec.repl.ReplExternalTables.Writer;
 
 public class ReplDumpTask extends Task<ReplDumpWork> implements Serializable {
+  private static final long serialVersionUID = 1L;
   private static final String dumpSchema = "dump_dir,last_repl_id#string,string";
   private static final String FUNCTION_METADATA_FILE_NAME = EximUtil.METADATA_NAME;
   private static final long SLEEP_TIME = 60000;
   Set<String> tablesForBootstrap = new HashSet<>();
+  private Path dumpAckFile;
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java]: This needn't be a member variable

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java (hunk 12 above)
Comment (pkumarsinha) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java]: Ack is being added at the end in these cases as well

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java (hunk 13 above)
Comment (pkumarsinha) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java]: We don't store the dumpath anywhere. Across the multiple call this info will be lost. I guess, ReplDumpWork is a better place for this info.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestScheduledReplicationScenarios.java [hunk 14]
```diff
@@ -151,6 +151,70 @@ public void testAcidTablesReplLoadBootstrapIncr() throws Throwable {
...
+            .run("create external table t1 (id int)")
+            .run("insert into t1 values(1)")
+            .run("insert into t1 values(2)");
+    try (ScheduledQueryExecutionService schqS =
+                 ScheduledQueryExecutionService.startScheduledQueryExecutorService(primary.hiveConf)) {
+      int next = 0;
+      ReplDumpWork.injectNextDumpDirForTest(String.valueOf(next));
+      primary.run("create scheduled query s1 every 10 minutes as repl dump " + primaryDbName);
+      primary.run("alter scheduled query s1 execute");
+      Thread.sleep(20000);
```
[the remark this answers is not in the document]
Comment (pkumarsinha) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestScheduledReplicationScenarios.java]: Ok

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java [hunk 15]
```diff
@@ -193,20 +193,29 @@ private Long getEventFromPreviousDumpMetadata(Path previousDumpPath) throws Sema
   }
 
   private Path getPreviousDumpMetadataPath(Path dumpRoot) throws IOException {
+    FileStatus latestUpdatedStatus = null;
     FileSystem fs = dumpRoot.getFileSystem(conf);
     if (fs.exists(dumpRoot)) {
       FileStatus[] statuses = fs.listStatus(dumpRoot);
       if (statuses.length > 0)  {
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java]: We don't need this check

Comment (pkumarsinha) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java]: Is fs.listStatus() guaranteed to be not null even when the the path does not have any folder/files?

Comment (aasha) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java]: It will return empty array

Comment (pkumarsinha) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java]: Done

File: ql/src/gen/thrift/gen-javabean/org/apache/hadoop/hive/ql/plan/api/StageType.java [hunk 16]
```diff
@@ -30,7 +30,7 @@
   REPL_TXN(15),
   REPL_INCREMENTAL_LOAD(16),
   SCHEDULED_QUERY_MAINT(17),
-  REPL_LOAD_COMPLETE_ACK(18);
+  REPL_OPERATION_COMPLETE_ACK(18);
 
   private final int value;
```
Comment (anishek) [inline ql/src/gen/thrift/gen-javabean/org/apache/hadoop/hive/ql/plan/api/StageType.java]: unused imports like HashMap, Util
remove the private qualifier from constructor.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java [hunk 17]
```diff
@@ -78,18 +81,21 @@
...
 import java.util.Base64;
-import java.util.ArrayList;
+import java.util.LinkedList;
 import java.util.UUID;
+import java.util.ArrayList;
 import java.util.concurrent.TimeUnit;
 import static org.apache.hadoop.hive.ql.exec.repl.ReplExternalTables.Writer;
 
 public class ReplDumpTask extends Task<ReplDumpWork> implements Serializable {
+  private static final long serialVersionUID = 1L;
```
Comment (anishek) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java]: tablesForBootstrap can be private

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java [hunk 18]
```diff
@@ -122,33 +128,38 @@ public String getName() {
...
-        Path hiveDumpRoot = new Path(currentDumpPath, ReplUtils.REPL_HIVE_BASE_DIR);
-        DumpMetaData dmd = new DumpMetaData(hiveDumpRoot, conf);
-        // Initialize ReplChangeManager instance since we will require it to encode file URI.
-        ReplChangeManager.getInstance(conf);
-        Path cmRoot = new Path(conf.getVar(HiveConf.ConfVars.REPLCMDIR));
-        Long lastReplId;
-        if (previousHiveDumpPath == null) {
-          lastReplId = bootStrapDump(hiveDumpRoot, dmd, cmRoot, hiveDb);
+      if (work.dirCopyIteratorInitialized() || work.replPathIteratorInitialized()) {
+        intitiateDataCopyTasks();
```
Comment (anishek) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java]: should be initiateDataCopyTasks

File: ql/src/test/org/apache/hadoop/hive/ql/exec/repl/TestReplDumpTask.java [hunk 19]
```diff
@@ -127,13 +128,15 @@ public void removeDBPropertyToPreventRenameWhenBootstrapDumpOfTableFails() throw
...
-                     long lastReplId, Hive hiveDb, HiveWrapper.Tuple<Table> tuple)
+      List<EximUtil.ReplPathMapping> dumpTable(String dbName, String tblName, String validTxnList, Path dbRoot,
+                                               Path replDataDir, long lastReplId, Hive hiveDb,
+                                               HiveWrapper.Tuple<Table> tuple)
           throws Exception {
         tableDumpCount++;
         if (tableDumpCount > 1) {
           throw new TestException();
         }
+        return Collections.EMPTY_LIST;
```
Comment (anishek) [inline ql/src/test/org/apache/hadoop/hive/ql/exec/repl/TestReplDumpTask.java]: better to use Collections.emptyList() here and other places, the generic cast would  be provided by this.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java [hunk 20]
```diff
@@ -662,15 +696,37 @@ void dumpTable(String dbName, String tblName, String validTxnList, Path dbRoot,
...
+    int maxTasks = conf.getIntVar(HiveConf.ConfVars.REPL_APPROX_MAX_LOAD_TASKS);
+    TaskTracker taskTracker = new TaskTracker(maxTasks);
+    while (taskTracker.canAddMoreTasks() && hasMoreCopyWork()) {
+      if (work.replPathIteratorInitialized() && extCopyWorkItr.hasNext()) {
+        childTasks.addAll(new ExternalTableCopyTaskBuilder(work, conf).tasks(taskTracker));
+      } else {
+        childTasks.addAll(ReplPathMapping.tasks(work, taskTracker, conf));
+      }
+    }
+    if (!childTasks.isEmpty()) {
```
Comment (anishek) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java]: better to not do the negate check and inverse the if else block by removing the !

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplOperationCompleteAckWork.java [hunk 21]
```diff
@@ -18,27 +18,28 @@
...
 /**
- * ReplLoadCompleteAckWork.
- * FS based Acknowledgement for repl load complete
+ * ReplOperationCompleteAckWork.
+ * FS based acknowledgement on repl dump and repl load completion.
  *
  */
-@Explain(displayName = "Repl Load Complete Ack", explainLevels = { Level.USER, Level.DEFAULT, Level.EXTENDED })
-public class ReplLoadCompleteAckWork implements Serializable {
+@Explain(displayName = "Repl Operation Complete Ack", explainLevels = { Level.USER, Level.DEFAULT, Level.EXTENDED })
```
Comment (anishek) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplOperationCompleteAckWork.java]: Just have the class as AckWork.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java [hunk 22]
```diff
@@ -662,15 +696,37 @@ void dumpTable(String dbName, String tblName, String validTxnList, Path dbRoot,
...
+    List<Task<?>> childTasks = new ArrayList<>();
+    int maxTasks = conf.getIntVar(HiveConf.ConfVars.REPL_APPROX_MAX_LOAD_TASKS);
+    TaskTracker taskTracker = new TaskTracker(maxTasks);
+    while (taskTracker.canAddMoreTasks() && hasMoreCopyWork()) {
+      if (work.replPathIteratorInitialized() && extCopyWorkItr.hasNext()) {
+        childTasks.addAll(new ExternalTableCopyTaskBuilder(work, conf).tasks(taskTracker));
+      } else {
+        childTasks.addAll(ReplPathMapping.tasks(work, taskTracker, conf));
+      }
+    }
```
Comment (anishek) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java]: why not just have just two statements like, with some return values added to child tasks.

    work.externalTableCopyTasks(taskTracker, conf);
    work.managedTableCopyTasks()

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java [hunk 23]
```diff
@@ -122,33 +128,38 @@ public String getName() {
...
   public int execute() {
     try {
-      Hive hiveDb = getHive();
-      Path dumpRoot = new Path(conf.getVar(HiveConf.ConfVars.REPLDIR),
-              Base64.getEncoder().encodeToString(work.dbNameOrPattern.toLowerCase()
-                      .getBytes(StandardCharsets.UTF_8.name())));
-      Path previousDumpMetaPath = getPreviousDumpMetadataPath(dumpRoot);
-      Path previousHiveDumpPath =
-              previousDumpMetaPath != null ? new Path(previousDumpMetaPath, ReplUtils.REPL_HIVE_BASE_DIR) : null;
-      //If no previous dump is present or previous dump was loaded, proceed
```
Comment (anishek) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java]: why not move the above statement in the function before  this

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java [hunk 24]
```diff
@@ -122,33 +128,38 @@ public String getName() {
...
         } else {
-          work.setEventFrom(getEventFromPreviousDumpMetadata(previousHiveDumpPath));
-          lastReplId = incrementalDump(hiveDumpRoot, dmd, cmRoot, hiveDb);
+          LOG.warn("Previous Dump is not yet loaded");
         }
-        prepareReturnValues(Arrays.asList(currentDumpPath.toUri().toString(), String.valueOf(lastReplId)));
-        writeDumpCompleteAck(hiveDumpRoot);
-        deleteAllPreviousDumpMeta(dumpRoot, currentDumpPath);
-      } else {
-        LOG.warn("Previous Dump is not yet loaded");
```
Comment (anishek) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java]: this is not warn its info

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java [hunk 25]
```diff
@@ -404,7 +417,8 @@ private Long incrementalDump(Path dumpRoot, DumpMetaData dmd, Path cmRoot, Hive
       }
 
       Path dbRoot = getBootstrapDbRoot(dumpRoot, dbName, true);
-
+      List<Path> extTableLocations = new LinkedList<>();
+      List<EximUtil.ReplPathMapping> replPathMappings = new ArrayList<>();
```
Comment (anishek) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java]: why is this repl path mapping ? isnt this managed table data only ?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java [hunk 26]
```diff
@@ -662,15 +696,37 @@ void dumpTable(String dbName, String tblName, String validTxnList, Path dbRoot,
...
+  private void intitiateDataCopyTasks() throws SemanticException {
+    Iterator<ExternalTableCopyTaskBuilder.DirCopyWork> extCopyWorkItr = work.getDirCopyIterator();
+    List<Task<?>> childTasks = new ArrayList<>();
+    int maxTasks = conf.getIntVar(HiveConf.ConfVars.REPL_APPROX_MAX_LOAD_TASKS);
+    TaskTracker taskTracker = new TaskTracker(maxTasks);
+    while (taskTracker.canAddMoreTasks() && hasMoreCopyWork()) {
+      if (work.replPathIteratorInitialized() && extCopyWorkItr.hasNext()) {
+        childTasks.addAll(new ExternalTableCopyTaskBuilder(work, conf).tasks(taskTracker));
+      } else {
+        childTasks.addAll(ReplPathMapping.tasks(work, taskTracker, conf));
```
Comment (anishek) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java]: why is this static method here, there is nothing accessed from there why not just in the current class. is some encapsulation broken here ?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java [hunk 27]
```diff
@@ -662,15 +696,37 @@ void dumpTable(String dbName, String tblName, String validTxnList, Path dbRoot,
...
+      }
+    }
+    if (!childTasks.isEmpty()) {
+      DAGTraversal.traverse(childTasks, new AddDependencyToLeaves(TaskFactory.get(work, conf)));
+    } else {
+      prepareReturnValues(work.getResultValues());
+      childTasks.add(TaskFactory.get(new ReplOperationCompleteAckWork(work.getDumpAckFile()), conf));
     }
```
Comment (anishek) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java]: why is ack work only if the child task is empty, there is possibility of adding the ack work at the end of all the dag, if the dag can contain more nodes. may be need to move it out of else clause and do it based on tracker.canAddMoreTasks or something?

Comment (pkumarsinha) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java]: Yes, in my previous patch, I was adding the ack work as child task whenever it was possible by making use of tracker.canAddMoreTasks . Then the requirement came to run this prepareReturnValues(); as the last operation just before ack task.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java (hunk 26 above)
Comment (pkumarsinha) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java]: ReplPathMapping.tasks() is sort of util method. Ideally there should be a builder class like we have for dirCopyTasks: ExternalTableCopyTaskBuilder.
In this case full fledged builder wasn't required. I can add a new Builder class for the same.
May be, we can shift it to ReplCopyTask as a static method as ultimately it is giving back ReplCopyTasks.

File: ql/src/test/org/apache/hadoop/hive/ql/exec/repl/TestReplDumpTask.java (hunk 19 above)
Comment (pkumarsinha) [reply ql/src/test/org/apache/hadoop/hive/ql/exec/repl/TestReplDumpTask.java]: done

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplOperationCompleteAckWork.java (hunk 21 above)
Comment (pkumarsinha) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplOperationCompleteAckWork.java]: Done

File: ql/src/gen/thrift/gen-javabean/org/apache/hadoop/hive/ql/plan/api/StageType.java (hunk 16 above)
Comment (pkumarsinha) [reply ql/src/gen/thrift/gen-javabean/org/apache/hadoop/hive/ql/plan/api/StageType.java]: Removed.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java (hunk 23 above)
Comment (pkumarsinha) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java]: Done

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java (hunk 24 above)
Comment (pkumarsinha) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java]: Done

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java (hunk 17 above)
Comment (pkumarsinha) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java]: done

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java (hunk 27 above)
Comment (pkumarsinha) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java]: Have refactored a bit.
